data/transactions/logic/opcodes.go

```diff
@@ -273,242 +343,225 @@ var anyIntInt = StackTypes{StackAny, StackUint64, StackUint64}
-	{0x28, "bytec_0", opByteConst0, asmDefault, disDefault, nil, oneBytes, 1, modeAny, opDefault},
-	{0x29, "bytec_1", opByteConst1, asmDefault, disDefault, nil, oneBytes, 1, modeAny, opDefault},
-	{0x2a, "bytec_2", opByteConst2, asmDefault, disDefault, nil, oneBytes, 1, modeAny, opDefault},
-	{0x2b, "bytec_3", opByteConst3, asmDefault, disDefault, nil, oneBytes, 1, modeAny, opDefault},
-	{0x2c, "arg", opArg, asmArg, disDefault, nil, oneBytes, 1, runModeSignature, immediates("n")},
-	{0x2d, "arg_0", opArg0, asmDefault, disDefault, nil, oneBytes, 1, runModeSignature, opDefault},
-	{0x2e, "arg_1", opArg1, asmDefault, disDefault, nil, oneBytes, 1, runModeSignature, opDefault},
-	{0x2f, "arg_2", opArg2, asmDefault, disDefault, nil, oneBytes, 1, runModeSignature, opDefault},
-	{0x30, "arg_3", opArg3, asmDefault, disDefault, nil, oneBytes, 1, runModeSignature, opDefault},
-	{0x31, "txn", opTxn, asmDefault, disDefault, nil, oneAny, 1, modeAny, field("f", &TxnScalarFields)},
+	{0x04, "ed25519verify", opEd25519Verify, proto("bbb:i"), 1, costly(1900).only(modeSig)},
+	{0x04, "ed25519verify", opEd25519Verify, proto("bbb:i"), 5, costly(1900)},
+
+	{0x05, "ecdsa_verify", opEcdsaVerify, proto("bbbbb:i"), 5, costByField("v", &EcdsaCurves, ecdsaVerifyCosts)},
+	{0x06, "ecdsa_pk_decompress", opEcdsaPkDecompress, proto("b:bb"), 5, costByField("v", &EcdsaCurves, ecdsaDecompressCosts)},
+	{0x07, "ecdsa_pk_recover", opEcdsaPkRecover, proto("bibb:bb"), 5, field("v", &EcdsaCurves).costs(2000)},
+
+	{0x08, "+", opPlus, proto("ii:i"), 1, opDefault()},
+	{0x09, "-", opMinus, proto("ii:i"), 1, opDefault()},
+	{0x0a, "/", opDiv, proto("ii:i"), 1, opDefault()},
+	{0x0b, "*", opMul, proto("ii:i"), 1, opDefault()},
+	{0x0c, "<", opLt, proto("ii:i"), 1, opDefault()},
+	{0x0d, ">", opGt, proto("ii:i"), 1, opDefault()},
+	{0x0e, "<=", opLe, proto("ii:i"), 1, opDefault()},
+	{0x0f, ">=", opGe, proto("ii:i"), 1, opDefault()},
+	{0x10, "&&", opAnd, proto("ii:i"), 1, opDefault()},
+	{0x11, "||", opOr, proto("ii:i"), 1, opDefault()},
+	{0x12, "==", opEq, proto("aa:i"), 1, stacky(typeEquals)},
+	{0x13, "!=", opNeq, proto("aa:i"), 1, stacky(typeEquals)},
+	{0x14, "!", opNot, proto("i:i"), 1, opDefault()},
+	{0x15, "len", opLen, proto("b:i"), 1, opDefault()},
+	{0x16, "itob", opItob, proto("i:b"), 1, opDefault()},
+	{0x17, "btoi", opBtoi, proto("b:i"), 1, opDefault()},
+	{0x18, "%", opModulo, proto("ii:i"), 1, opDefault()},
+	{0x19, "|", opBitOr, proto("ii:i"), 1, opDefault()},
+	{0x1a, "&", opBitAnd, proto("ii:i"), 1, opDefault()},
+	{0x1b, "^", opBitXor, proto("ii:i"), 1, opDefault()},
+	{0x1c, "~", opBitNot, proto("i:i"), 1, opDefault()},
+	{0x1d, "mulw", opMulw, proto("ii:ii"), 1, opDefault()},
+	{0x1e, "addw", opAddw, proto("ii:ii"), 2, opDefault()},
+	{0x1f, "divmodw", opDivModw, proto("iiii:iiii"), 4, costly(20)},
+
+	{0x20, "intcblock", opIntConstBlock, proto(":"), 1, constants(asmIntCBlock, checkIntConstBlock, "uint ...", immInts)},
```

Comment:
It doesn't predate the PR, at least not by much if it does.  As part of handling deadcode, I wanted the OpSpec to clearly indicate which functions always exit.  Code that follows them is dead, and need not type check.  So `nil` simply means an opcode puts nothing on the stack. `StackNone` says the opcode doesn't even return (mnemonic "eXit" is why I used "x" for the character).

Reply:
@jannotti Got it - Thanks for the explanation.

What do you think of renaming `StackNone` to something that more directly hints at an unconditional termination (e.g. `StackExit`)?

I originally interpreted `nil` semantics to be `StackNone` because of the _none_, and I hadn't yet realized `nil` is an option.   While I think the naming can be easier to interpret, I don't feel strongly.  If you prefer to keep as is, then i'm good to resolve the thread.

Reply:
I'll think it through.  StackNone has some other purposes, and StackNone being at the top of the stack doesn't _only_ happen because of opcodes that exit. It also happens from unconditional jumps, `b` and `callsub` (though it is immediately removed in the case of `callsub` because `callsub` _also_ acts as an incoming "label" point.).  assembly adds StackNone if `spec.alwaysJumps()`.

Perhaps adding a type, StackDead, for use by the exiting and jumping instructions would be clearest.

Reply:
@jannotti Sounds good to at least consider the topic.  Generally, I _imagine_ I'd favor further specificity (e.g. `StackDead`).  I'll leave the thread open while the topic is considered.

Reply:
I think the right thing to do is to have a flag on ProgramKnowledge, instead of encoding deadness as a sepcial type at the tip. I did that before ProgramKnowledge existed, as a way to wedge in deadness state to the existing model which was only a type stack.  I'll probably change this, unless I find out it's more involved than I think.  Give me a couple hours.

Reply:
Ok, I simplified `ProgramKnowledge` so it no longer uses StackNone at the end of the stack to mean deadcode, now it's explicit.

I still use StackNone in the Return.Types of an OpSpec (by using "x") to indicate that the opcode always exits though. I suppose I could have AlwaysExits() simply _know_ that `err` and `return` are the only examples, or add another field to OpSpec to indicate an opcode always exits. But I think I prefer this. It feels reasonably natural to mark this property as a special form of indicating what the opcode returns.

Reply:
@jannotti https://github.com/algorand/go-algorand/pull/3870/commits/caca024a29873d92a66f08c6667685c7469aca98 works for me.  Resolving.